### ISSUE TITLE
Fails on git source uninstalled even though its group is excluded

### ISSUE THREAD

**Issue by seanlinsley** [bundler/bundler#4699]
Note that the gem in question should be excluded because the groups it's contained in are excluded. And yet, `bundle install` fails because it's not installed.

We just discovered this while doing a deploy, which installs the latest version of Bundler (1.12.5), but I was also able to reproduce locally (1.11.2).

Here's an abbreviated version of the commands I ran locally to reproduce, and isolate the cause:

``` sh
$ bundle
Bundle complete! 89 Gemfile dependencies, 229 gems now installed.

$ rm -rf `bundle show jasmine-rails`

$ bundle install --quiet --without development test
https://github.com/seanlinsley/jasmine-rails.git (at phantomjs2@64872fd) is not yet checked out. Run `bundle install` first.

$ bundle install --quiet
https://github.com/seanlinsley/jasmine-rails.git (at phantomjs2@64872fd) is not yet checked out. Run `bundle install` first.

$ bundle install
https://github.com/seanlinsley/jasmine-rails.git (at phantomjs2@64872fd) is not yet checked out. Run `bundle install` first.

$ bundle --version
Bundler version 1.11.2

$ bundle config
Settings are listed in order of priority. The top value will be used.
with
Set for your local app (.bundle/config): "pry"
Set for the current user (~/.bundle/config): "pry"

without
Set for your local app (.bundle/config): "development:test"

$ rm -rf .bundle

$ bundle
Bundle complete! 89 Gemfile dependencies, 229 gems now installed.

$ rm -rf `bundle show jasmine-rails`

$ bundle config
Settings are listed in order of priority. The top value will be used.
with
Set for your local app (.bundle/config): "pry"
Set for the current user (~/.bundle/config): "pry"

$ bundle --without development
Bundle complete! 89 Gemfile dependencies, 204 gems now installed.

$ rm -rf `bundle show jasmine-rails`

$ bundle install --without development test
https://github.com/seanlinsley/jasmine-rails.git (at phantomjs2@64872fd) is not yet checked out. Run `bundle install` first.

$ bundle install --without development
https://github.com/seanlinsley/jasmine-rails.git (at phantomjs2@64872fd) is not yet checked out. Run `bundle install` first.

$ bundle config
Settings are listed in order of priority. The top value will be used.
with
Set for your local app (.bundle/config): "pry"
Set for the current user (~/.bundle/config): "pry"

without
Set for your local app (.bundle/config): "development:test"
```

These are the relevant parts of the Gemfile:

``` ruby
# Force the `github:` syntax to use HTTPS instead of the Git protocol.
# The Git protocol isn't encrypted, and so is subject to MiTM attacks.
# This can be removed once Bundler 2.0 is released.
# https://github.com/bundler/bundler/pull/2569
git_source :github do |repo|
  repo = "#{repo}/#{repo}" unless repo.include? '/'
  "https://github.com/#{repo}.git"
end

group :development, :test do
  # Debugging
  gem 'jasmine-rails', github: 'seanlinsley/jasmine-rails', branch: 'phantomjs2'
end
```

**Comment by seanlinsley** [comment]
Here's the verbose output (with the `.bundle/config` already set):
```
$ bundle --verbose
Bundler::PathError: The path `bundler/gems/jasmine-rails-64872fd078b5` does not exist.
gems/bundler-1.11.2/lib/bundler/source/path.rb:160:in `load_spec_files'
  gems/bundler-1.11.2/lib/bundler/source/git.rb:188:in `load_spec_files'
  gems/bundler-1.11.2/lib/bundler/source/path.rb:92:in `local_specs'
  gems/bundler-1.11.2/lib/bundler/source/git.rb:159:in `specs'
  gems/bundler-1.11.2/lib/bundler/definition.rb:209:in `block (2 levels) in index'
  gems/bundler-1.11.2/lib/bundler/definition.rb:207:in `each'
  gems/bundler-1.11.2/lib/bundler/definition.rb:207:in `block in index'
  gems/bundler-1.11.2/lib/bundler/index.rb:9:in `build'
  gems/bundler-1.11.2/lib/bundler/definition.rb:204:in `index'
  gems/bundler-1.11.2/lib/bundler/definition.rb:140:in `specs'
  gems/bundler-1.11.2/lib/bundler/environment.rb:22:in `specs'
  gems/bundler-1.11.2/lib/bundler/installer.rb:170:in `install_in_parallel'
  gems/bundler-1.11.2/lib/bundler/installer.rb:155:in `install'
  gems/bundler-1.11.2/lib/bundler/installer.rb:70:in `run'
  gems/bundler-1.11.2/lib/bundler/installer.rb:21:in `install'
  gems/bundler-1.11.2/lib/bundler/cli/install.rb:105:in `run'
  gems/bundler-1.11.2/lib/bundler/cli.rb:172:in `install'
  gems/bundler-1.11.2/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  gems/bundler-1.11.2/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
  gems/bundler-1.11.2/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
  gems/bundler-1.11.2/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
  gems/bundler-1.11.2/lib/bundler/cli.rb:10:in `start'
  gems/bundler-1.11.2/exe/bundle:19:in `block in <top (required)>'
  gems/bundler-1.11.2/lib/bundler/friendly_errors.rb:7:in `with_friendly_errors'
  gems/bundler-1.11.2/exe/bundle:17:in `<top (required)>'
  bin/bundle:23:in `load'
  bin/bundle:23:in `<main>'
  bin/ruby_executable_hooks:15:in `eval'
  bin/ruby_executable_hooks:15:in `<main>'
https://github.com/seanlinsley/jasmine-rails.git (at phantomjs2@64872fd) is not yet checked out. Run `bundle install` first.

**Comment by segiddins** [comment]
Can you check the 1.13 prerelease? I seem to recall us spending time on this issue recently. Thanks!

**Comment by seanlinsley** [comment]
I'm having trouble installing the prerelease version:
```
$ gem install bundler --pre

Fetching: bundler-1.13.0.pre.1.gem (100%)
Successfully installed bundler-1.13.0.pre.1
Parsing documentation for bundler-1.13.0.pre.1
Installing ri documentation for bundler-1.13.0.pre.1
Done installing documentation for bundler after 4 seconds
1 gem installed

$ gem uninstall bundler

Select gem to uninstall:
 1. bundler-1.11.2
 2. bundler-1.13.0.pre.1
 3. All versions
> 1
Successfully uninstalled bundler-1.11.2

$ bundle --version

Fetching: bundler-1.11.2.gem (100%)
Bundler version 1.11.2

**Comment by seanlinsley** [comment]
Oh looks like it was looking up the Bundler version from `Gemfile.lock`. Manually editing that file got it to use the prerelease version.

Is there a better way of doing that?

**Comment by seanlinsley** [comment]
It looks like the prerelease version does indeed fix this issue 💖

**Comment by segiddins** [comment]
@seanlinsley `bundle update --bundler`

**Comment by seanlinsley** [comment]
Is that documented somewhere? For example I looked at `CONTRIBUTING.md` and it didn't mention it:

<img width="684" alt="screen shot 2016-06-22 at 11 15 33 am" src="https://cloud.githubusercontent.com/assets/688886/16274331/c32701ac-386a-11e6-91ed-b25c394a70f6.png">

**Comment by seanlinsley** [comment]
I'm only leaving this open as an opportunity to improve the documentation around using prerelease versions.

**Comment by segiddins** [comment]
@seanlinsley it's in the man page for bundle update, and is about a new feature in 1.13 (bundler version trampolining) rather than this being a prerelease

**Comment by indirect** [comment]
@seanlinsley thanks for pointing that out!

@segiddins ooh... we need to update the documentation for using prereleases now that the 1.13 prerelease will automatically trampoline to older versions 😅

**Comment by segiddins** [comment]
This was reported fixed in 1.13, so closing.
